pyscf/adc/radc.py

```diff
@@ -0,0 +1,1751 @@
+
+    eris_ovov = eris.ovov
+
+    t2_1  = t2[0]
+
+    #Compute MP2 correlation energy
+
+    e_mp2 = 0.5 * np.einsum('ijab,iajb', t2_1, eris_ovov)
+    e_mp2 -= 0.5 * np.einsum('ijab,ibja', t2_1, eris_ovov)
+    e_mp2 -= 0.5 * np.einsum('jiab,iajb', t2_1, eris_ovov)
+    e_mp2 += 0.5 * np.einsum('jiab,ibja', t2_1, eris_ovov)
+    e_mp2 += np.einsum('ijab,iajb', t2_1, eris_ovov)
+
+    e_corr = e_mp2
+
+    if (myadc.method == "adc(3)"):
+
+        t2_1_a = t2_1 - t2_1.transpose(1,0,2,3).copy()
+
+        #Compute MP3 correlation energy
+        eris_oovv = eris.oovv
+        eris_ovvo = eris.ovvo
+        eris_oooo = eris.oooo
+
+        temp = t2_1_a.reshape(nocc*nocc,nvir*nvir)
+        eris_vvvv = radc_ao2mo.unpack_eri_2(eris.vvvv, nvir)
+        eris_vvvv = eris_vvvv.transpose(0,2,1,3)
+        eris_vvvv = eris_vvvv.copy()[:].reshape(nvir*nvir,nvir*nvir)
+        temp_1_a = np.dot(temp,eris_vvvv.T).reshape(nocc,nocc,nvir,nvir)
+    
+        temp_2 = t2_1.reshape(nocc*nocc,nvir*nvir)
+        temp_1_ab = np.dot(temp_2,eris_vvvv.T).reshape(nocc,nocc,nvir,nvir)
+        e_mp3 =  np.einsum('ijcd,ijcd',temp_1_ab, t2_1)
+        del temp_1_ab
+
+        eris_vvvv = eris_vvvv[:].reshape(nvir,nvir,nvir,nvir)
+        eris_vvvv = eris_vvvv.transpose(0,1,3,2)
+        eris_vvvv = eris_vvvv.copy()[:].reshape(nvir*nvir,nvir*nvir)
+        temp_1_a -= np.dot(temp,eris_vvvv.T).reshape(nocc,nocc,nvir,nvir)
+        del eris_vvvv
+        e_mp3 += 0.25 * np.einsum('ijcd,ijcd',temp_1_a, t2_1_a)
+        del temp_1_a
+        
+        temp_2_a =  np.einsum('ijab,klab', t2_1_a, t2_1_a)
+        e_mp3 += 0.25 * np.einsum('ijkl,ikjl',temp_2_a, eris_oooo)
+        e_mp3 -= 0.25 * np.einsum('ijkl,iljk',temp_2_a, eris_oooo)
+        del temp_2_a
+
+        temp_2 =  np.einsum('ijab,klab', t2_1, t2_1)
+        e_mp3 +=  np.einsum('ijkl,ikjl',temp_2, eris_oooo)
+        del temp_2
+
+        temp_3_a = np.einsum('ijab,ikcb->akcj', t2_1_a, t2_1_a)
+        temp_3_a += np.einsum('jiab,kicb->akcj', t2_1, t2_1)
+        e_mp3 -= 2 * np.einsum('akcj,kjac',temp_3_a, eris_oovv)
+        e_mp3 += 2 * np.einsum('akcj,kcaj',temp_3_a, eris_ovvo)
+        del temp_3_a
+
+        temp_3 = np.einsum('ijab,ikcb->akcj', t2_1, t2_1)
+        e_mp3 -= np.einsum('akcj,kjac',temp_3, eris_oovv)
+        del temp_3
+   
+        temp_3_1 = np.einsum('jiba,kibc->akcj', t2_1, t2_1)
+        e_mp3 -= np.einsum('akcj,kjac',temp_3_1, eris_oovv)
+        del temp_3_1
+
+        temp_3_2 = -np.einsum('ijab,ikbc->akcj', t2_1_a, t2_1)
+        temp_3_2 -= np.einsum('jiab,ikcb->akcj', t2_1, t2_1_a)
+        e_mp3 += np.einsum('akcj,kcaj',temp_3_2, eris_ovvo)
+        del temp_3_2
+
+        temp_3_3 = -np.einsum('ijba,ikcb->akcj', t2_1, t2_1_a)
+        temp_3_3 -= np.einsum('ijab,kicb->akcj', t2_1_a, t2_1)
+        e_mp3 += np.einsum('akcj,kcaj',temp_3_3, eris_ovvo)
+        del temp_3_3
+    
+        e_corr += e_mp3
+
+    return e_corr
+
+
+class RADC(lib.StreamObject):
```

Comment:
We may want to create an ADC class with common functions and derive RADC and UADC classes from that one